RecoCTPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc

```diff
@@ -230,66 +256,146 @@ reco::ForwardProton ProtonReconstructionAlgorithm::reconstructFromMultiRP(const
        << "    initial estimate: xi_init = " << xi_init << ", th_x_init = " << th_x_init
        << ", th_y_init = " << th_y_init << ", vtx_y_init = " << vtx_y_init << "." << std::endl;
 
-  // minimisation
-  fitter_->Config().ParSettings(0).Set("xi", xi_init, 0.005);
-  fitter_->Config().ParSettings(1).Set("th_x", th_x_init, 2E-6);
-  fitter_->Config().ParSettings(2).Set("th_y", th_y_init, 2E-6);
-  fitter_->Config().ParSettings(3).Set("vtx_y", vtx_y_init, 10E-6);
+  // prepare result containers
+  bool valid = false;
+  double chi2 = 0.;
+  double xi = 0., th_x = 0., th_y = 0., vtx_y = 0.;
 
-  if (!fitVtxY_)
-    fitter_->Config().ParSettings(3).Fix();
+  using FP = reco::ForwardProton;
+  FP::CovarianceMatrix cm;
 
-  chiSquareCalculator_->tracks = &tracks;
-  chiSquareCalculator_->m_rp_optics = &m_rp_optics_;
+  if (multi_rp_algorithm_ == mrpaChi2) {
+    // minimisation
+    fitter_->Config().ParSettings(0).Set("xi", xi_init, 0.005);
+    fitter_->Config().ParSettings(1).Set("th_x", th_x_init, 2E-6);
+    fitter_->Config().ParSettings(2).Set("th_y", th_y_init, 2E-6);
+    fitter_->Config().ParSettings(3).Set("vtx_y", vtx_y_init, 10E-6);
+
+    if (!fitVtxY_)
+      fitter_->Config().ParSettings(3).Fix();
+
+    chiSquareCalculator_->tracks = &tracks;
+    chiSquareCalculator_->m_rp_optics = &m_rp_optics_;
+
+    fitter_->FitFCN();
+    fitter_->FitFCN();  // second minimisation in case the first one had troubles
+
+    // extract proton parameters
+    const ROOT::Fit::FitResult &result = fitter_->Result();
+    const double *params = result.GetParams();
+
+    valid = result.IsValid();
+    chi2 = result.Chi2();
+    xi = params[0];
+    th_x = params[1];
+    th_y = params[2];
+    vtx_y = params[3];
+
+    map<unsigned int, signed int> index_map = {
+        {(unsigned int)FP::Index::xi, 0},
+        {(unsigned int)FP::Index::th_x, 1},
+        {(unsigned int)FP::Index::th_y, 2},
+        {(unsigned int)FP::Index::vtx_y, ((fitVtxY_) ? 3 : -1)},
+        {(unsigned int)FP::Index::vtx_x, -1},
+    };
+
+    for (unsigned int i = 0; i < (unsigned int)FP::Index::num_indices; ++i) {
+      signed int fit_i = index_map[i];
+
+      for (unsigned int j = 0; j < (unsigned int)FP::Index::num_indices; ++j) {
+        signed int fit_j = index_map[j];
+
+        cm(i, j) = (fit_i >= 0 && fit_j >= 0) ? result.CovMatrix(fit_i, fit_j) : 0.;
+      }
+    }
+  }
 
-  fitter_->FitFCN();
-  fitter_->FitFCN();  // second minimisation in case the first one had troubles
+  if (multi_rp_algorithm_ == mrpaNewton || multi_rp_algorithm_ == mrpaAnalIter) {
```

Comment:
```else if ```

Also, it may be time to consider making separate methods to handle each case, especially if any more developments are coming in.

Reply:
I did consider it. However, as some parts of the code are shared between different algorithms, I decided to put everything in the same method. No further development is foreseen in this moment. Should this change I agree that we should probably factorise the code into smaller methods.